crates/accelerate/src/inverse_cancellation.rs

```diff
@@ -0,0 +1,193 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use ahash::RandomState;
+use hashbrown::HashSet;
+use indexmap::IndexMap;
+use pyo3::prelude::*;
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+use qiskit_circuit::circuit_instruction::OperationFromPython;
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::operations::Operation;
+use qiskit_circuit::packed_instruction::PackedInstruction;
+
+fn gate_eq(py: Python, gate_a: &PackedInstruction, gate_b: &OperationFromPython) -> PyResult<bool> {
+    if gate_a.op.name() != gate_b.operation.name() {
+        return Ok(false);
+    }
+    let a_params = gate_a.params_view();
+    if a_params.len() != gate_b.params.len() {
+        return Ok(false);
+    }
+    let mut param_eq = true;
+    for (a, b) in a_params.iter().zip(&gate_b.params) {
+        if !a.is_close(py, b, 1e-10)? {
+            param_eq = false;
+            break;
+        }
+    }
+    Ok(param_eq)
+}
+
+fn run_on_self_inverse(
+    py: Python,
+    dag: &mut DAGCircuit,
+    op_counts: &IndexMap<String, usize, RandomState>,
+    self_inverse_gate_names: HashSet<String>,
+    self_inverse_gates: Vec<OperationFromPython>,
+) -> PyResult<()> {
+    if !self_inverse_gate_names
+        .iter()
+        .any(|name| op_counts.contains_key(name))
+    {
+        return Ok(());
+    }
+    for gate in self_inverse_gates {
+        let gate_count = op_counts.get(gate.operation.name()).unwrap_or(&0);
+        if *gate_count <= 1 {
+            continue;
+        }
+        let mut collect_set: HashSet<String> = HashSet::with_capacity(1);
+        collect_set.insert(gate.operation.name().to_string());
+        let gate_runs: Vec<Vec<NodeIndex>> = dag.collect_runs(collect_set).unwrap().collect();
+        for gate_cancel_run in gate_runs {
+            let mut partitions: Vec<Vec<NodeIndex>> = Vec::new();
+            let mut chunk: Vec<NodeIndex> = Vec::new();
+            let max_index = gate_cancel_run.len() - 1;
+            for (i, cancel_gate) in gate_cancel_run.iter().enumerate() {
+                let node = &dag.dag[*cancel_gate];
+                if let NodeType::Operation(inst) = node {
+                    if gate_eq(py, inst, &gate)? {
+                        chunk.push(*cancel_gate);
+                    } else {
+                        let is_empty: bool = chunk.is_empty();
```

Comment:
IMO having the temporary variable here hampers readability a tad.

Reply:
Removed in: https://github.com/Qiskit/qiskit/pull/13013/commits/efeeda5534e7bf778547674d77cdb819409e1c83